server/src/main/scala/viper/gobraserver/GobraServer.scala

```diff
@@ -55,56 +58,63 @@ object GobraServer extends GobraFrontend {
     VerifierState.flushCachedDiagnostics()
   }
 
-  def restart(): Unit = {
+  def restart(): Future[Unit] = {
     stop()
-    delete()
-    val options = _options
-    val executor = _executor
-    init(options)(executor)
-    start()
+      .flatMap(_ => {
+        delete()
+        _executor.restart()
+      })(_executor)
+      .map(_ => {
+        val options = _options
+        val executor = _executor
+        init(options)(executor)
+        start()
+      })(_executor)
   }
 
   private def serverExceptionHandling(fileData: FileData, resultFuture: Future[VerifierResult])(implicit executor: GobraExecutionContext): Future[VerifierResult] = {
 
     val fileUri = fileData.fileUri
 
     // do some post processing if verification has failed
-    resultFuture.recoverWith({ case exception =>
-      // restart Gobra Server and then update client state
-      restart()
-
-      exception match {
-        case e: Violation.LogicException =>
-          VerifierState.removeDiagnostics(fileUri)
-          val overallResult = Helper.getOverallVerificationResultFromException(fileUri, e)
-
-          VerifierState.updateVerificationInformation(fileUri, Right(overallResult))
-
-
-          if (fileUri == VerifierState.openFileUri) {
-            VerifierState.publishDiagnostics(fileUri)
-          }
-
-        case e =>
-          println("Exception occurred:")
-          e.printStackTrace()
-
-          // remove verification information about this file
-          // otherwise, reopening this file in the client will result in sending the last progress although no
-          // verification is running
-          VerifierState.removeVerificationInformation(fileUri)
-
-          VerifierState.client match {
-            case Some(c) =>
-              c.showMessage(new MessageParams(MessageType.Error, "An exception occurred during verification: " + e))
-              c.verificationException(fileUri)
-            case None =>
+    resultFuture.transformWith {
+      case Success(res) => Future.successful(res)
+      case Failure(exception: Throwable) =>
```

Comment:
Why do you need the type annotation. Is there a case where a failure does not contain a Throwable? How is this case handled in the code?

Reply:
No I don't need it